jfix-zookeeper/src/main/java/ru/fix/zookeeper/lock/PersistentExpiringLockManager.java

```diff
@@ -148,36 +138,155 @@ public void release(LockIdentity lockId) {
         }
     }
 
+    @Override
+    public void close() {
+        locksContainer.close();
+        lockProlongationScheduler.close();
+    }
+
     /**
-     * @param lockId
-     * @param lock
-     * @return false in case of Exception of lock expiration or losing lock ownership
+     * Contains the locks and provide synced access to remove, prolong and getState operations under locks
      */
-    private boolean checkAndProlongLockIfRequired(LockIdentity lockId, PersistentExpiringDistributedLock lock) {
-        try {
-            logger.debug("Check and prolong lockId={}", lockId);
-            return lock.checkAndProlongIfExpiresIn(
-                    config.get().getLockAcquirePeriod(),
-                    config.get().getExpirationPeriod());
-        } catch (Exception e) {
-            logger.error("Failed to checkAndProlongIfExpiresIn persistent locks with lockId {}", lockId, e);
-            return false;
+    static class ActiveLocksContainer implements AutoCloseable {
+        /**
+         * remove, prolong and getState operations MUST BE synced under {@link #globalLock}
+         */
+        private final ConcurrentMap<LockIdentity, LockContainer> locks = new ConcurrentHashMap<>();
+        private final Object globalLock = new Object();
+
+        public void putLock(
+                LockIdentity lockId,
+                PersistentExpiringDistributedLock lock,
+                LockProlongationFailedListener listener
+        ) {
+            LockContainer newLockContainer = new LockContainer(lock, listener);
+            LockContainer oldLockContainer = locks.put(lockId, newLockContainer);
```

Comment:
Почему бы нам не использовать putIfAbsent? т.е. мы видим, внешнение приложение пытается создать еще один лок с таким же ключом, мы создаем новое значение LockContainer и ругаемся в лог, а почему бы не вернуть текущее значение, сравнить и так же записать в log.error? 
Нам это дает следующее значение LockContainer в мапе остается неизменным, тогда вместо globallock мы можем использовать synchronized по значению:

```
            LockContainer lockContainer = locks.get(lockId);
            synchronized (lockContainer) {
```

Reply:
Synchronization on `lockContainer` instead of globalLock sounds more fine grained, but it will cost us rising complexity of the code. Since there will be no noticeable performance gain, I suggest not to jump into premature optimization.

Reply:
LockContainer is encapsulated inside ActiveLocksContainer so I think if we save this state we could easily synchronize with it (LockContainer) 
Added note